sql: Allow Alter Column Drop Identity Command

delphamk opened the pull request:
Informs: #110010

This change adds a new command, 'ALTER TABLE
ALTER [COLUMN] DROP IDENTITY [ IF EXISTS ]',
which allows an ID column to drop the related
its Identity constraint and related sequence.
'IF EXISTS' is optional and will provide a notice
if the column is not an identity.

Release note (sql change): Identity Columns can
drop the Identity constraint by running 'ALTER
TABLE t ALTER COLUMN c DROP IDENTITY [ IF EXISTS ]'.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119263)
<!-- Reviewable:end -->

delphamk committed:
sql: Allow Alter Column Drop Identity Command

Informs: #110010

This change adds a new command, 'ALTER TABLE
ALTER [COLUMN] DROP IDENTITY [ IF EXISTS ]',
which allows an IDENTITY column to drop its
Identity constraint and related sequence.
'IF EXISTS' is optional and will provide a notice
if the column is not an identity.

Release note (sql change): Identity Columns can
drop the Identity constraint and related sequence by running:
-'ALTER TABLE t ALTER COLUMN c DROP IDENTITY'.

Option 'IF EXISTS' can be added to skip the command
if the column in not an identity:
-'ALTER TABLE t ALTER COLUMN c DROP IDENTITY [ IF EXISTS ]'.

andrew-delph commented:
I think this got closed by mistake when syncing my fork!

andrew-delph commented on pkg/sql/drop_sequence.go:
```diff
@@ -190,7 +189,7 @@ func (p *planner) canRemoveOwnedSequencesImpl(
 		multipleIterationErr := seqDesc.ForeachDependedOnBy(func(dep *descpb.TableDescriptor_Reference) error {
 			if firstDep != nil {
-				return iterutil.StopIteration()
+				return errors.Newf("multiple iterations")
```
If StopIteration is returned then the error will be nil.

This also fixes an error which can be tested with
```

statement ok
CREATE TABLE test1 (a int GENERATED ALWAYS AS IDENTITY (START WITH 10));

statement ok
CREATE TABLE test2(id INT PRIMARY KEY DEFAULT nextval('public.test1_a_seq'));

statement ok
SET use_declarative_schema_changer = off

# this should error
statement ok
drop table test1;

statement error descriptor is being dropped
INSERT INTO test2 DEFAULT VALUES;
```

annrpom reviewed (approved):
great job and thanks for your contribution! just fixed up some typos and added pgcodes to the `statement error`s

(edit: i am merging this now)

bors r+

andrew-delph commented:
Thank you!
